src/wallet/asyncrpcoperation_sendmany.cpp

```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
+                shieldedRecipients += 1;
+                if (isfromsprout_ && !allowRevealedAmounts_) {
+                    throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
+                            "Sending from Sprout to Sapling is not enabled by default because it will "
+                            "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                            "Resubmit with the `allowRevealedAmounts` parameter set to `true` if "
+                            "you wish to allow this transaction to proceed anyway.");
+                }
+            },
+            [&](const libzcash::UnifiedAddress& ua) {
+                // unreachable; currently disallowed by checks at construction
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Sending to unified addresses is disabled.");
+            }
+        }, recipient.address);
     }
 
     CAmount sendAmount = txValues.z_outputs_total + txValues.t_outputs_total;
-    txValues.targetAmount = sendAmount + minersFee;
-
-    // When spending coinbase utxos, you can only specify a single zaddr as the change must go somewhere
-    // and if there are multiple zaddrs, we don't know where to send it.
-    if (isfromtaddr_) {
-        // Only select coinbase if we are spending from a single t-address to a single z-address.
-        if (!useanyutxo_ && isSingleZaddrOutput) {
-            bool b = find_utxos(true, txValues);
-            if (!b) {
-                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient transparent funds, no UTXOs found for taddr from address.");
-            }
+    txValues.targetAmount = sendAmount + fee_;
+
+    builder_.SetFee(fee_);
+
+    // Only select coinbase if we are spending from at most a single t-address.
+    bool allowTransparentCoinbase =
+        !IsFromAnyTaddr(paymentSource_) && // allow coinbase inputs from at most a single t-addr
+        transparentRecipients == 0; // cannot send transparent coinbase to transparent recipients
+
+    // Set the dust threshold so that we can select enough inputs to avoid
+    // creating dust change amounts.
+    CAmount dustThreshold{DefaultDustThreshold()};
+
+    // Find spendable inputs, and select a minimal set of them that
+    // can supply the required target amount.
+    auto spendable = FindSpendableInputs(allowTransparentCoinbase);
+    if (!spendable.LimitToAmount(txValues.targetAmount, dustThreshold)) {
+        CAmount changeAmount{spendable.Total() - txValues.targetAmount};
+        if (changeAmount > 0 && changeAmount < dustThreshold) {
+            // TODO: this condition is silly; we should provide the option for the caller to
+            // forego change (definitionally amount below the dust amount) and send the
+            // extra to the recipient or the miner fee to avoid creating dust change, rather
+            // than prohibit them from sending entirely in this circumstance.
```

Comment:
For dust shielded notes, that potentially leaks information about the note value, so we should do this carefully.

Reply:
Suppose that you're in this situation: you have a 100 ZEC note, and your counterparty requests a 99.99998999 ZEC payment. It's a little odd that that z_sendmany will not currently permit you to pay this bill unless you explicitly send 99.99999000. One alternative is to include this suggestion in the error message; then the user can explicitly opt-in to it by recreating their transaction, but that's just a hassle for the user when we could provide a flag to allow them to ask the machine to do it.

Reply:
Yes, but we should not make fingerprinting attacks easier.

Reply:
The real solution to dust issues is #2133 :-p